src/main/scala/org/scalacheck/Gen.scala

```diff
@@ -520,15 +520,21 @@ object Gen extends GenArities{
   /** Creates a resized version of a generator */
   def resize[T](s: Int, g: Gen[T]) = gen((p, seed) => g.doApply(p.withSize(s), seed))
 
-  /** Picks a random value from a list */
-  def oneOf[T](xs: Seq[T]): Gen[T] =
+  /** Picks a random value from a list. */
+  def oneOf[T](xs: Iterable[T]): Gen[T] =
     if (xs.isEmpty) {
       throw new IllegalArgumentException("oneOf called on empty collection")
     } else {
       val vector = xs.toVector
       choose(0, vector.size - 1).map(vector(_))
     }
 
+  /** Picks a random value from a list.
+   *  @todo Remove this overloaded method in the next major release. See #438.
+   */
+  def oneOf[T](xs: Seq[T]): Gen[T] =
+    oneOf(xs.asInstanceOf[Iterable[T]])
```

Comment:
Is the cast necessary?  I assumed that it would just be `oneOf(xs)`.

Reply:
Yes, a form of static casting is necessary. Without the cast, the following example would lead to an infinite loop.
```scala
val xs: Seq[Int] = ???
Gen.oneOf(xs)
```

For background, Scala (and Java) dispatches methods dynamically based on the run-time type of the receiver, i.e. `this`, and then statically based on the compile-time type of the arguments, e.g. `xs`. Hence, a recursive overloaded method with the same arguments (without casting) should always dispatch to itself.

For static casting, I believe that an alternative syntax (unique to Scala) is type ascription, which has the form
```scala
oneOf(xs: Iterable[T])
```
However, I am uncertain of how it would be behave differently than `asInstanceOf`; I  can look into this later tonight.

Reply:
After reading into the subject, it seems that the type ascription will provide a compile-time hint to the compiler to dispatch the appropriate overloaded method. Since this is a minor run-time improvement and likely better Scala style, I will proceed with this unless there is a better option.

Thanks for bringing this to my attention @ashawley . :)

Reply:
The only other option would be doing `oneOf(xs.toIterable)`.  It sounds like the type annotation is best and does the right thing.  Thanks.

Reply:
I looked at the results of deconstructing the class file with `javap` and it seems the type annotation is the equivalent of using `asInstanceOf`, and `toIterable` gives an unnecessary extra call, so really it's a question of readability between the type annotation and `asInstanceOf`.

Reply:
> type annotation is the equivalent of using asInstanceOf

Sorry to butt in, I'm afraid that's only true of the final byte code, there is a very important difference at compile time. The type annotation is safer in that it has a compile time check, the `asInstanceOf` throws all type safety out the window:

```scala
> "" : Int
<console>:8: error: type mismatch;
 found   : String("")
 required: Int
              "" : Int

> "".asInstanceOf[Int]
java.lang.ClassCastException: java.lang.String cannot be cast to java.lang.Integer
```

As you can see, `String` is definitely not `Int`, but you can force the conversation anyway with `asInstanceOf` and it will quite happily compile (but may fail at runtime).

In terms of the code in question, you should absolutely use a type annotation. `asInstanceOf` will continue to work even if `xs` later changes to something that's not actually an `Iterable`.

Reply:
Thanks for verifying the assumptions @ashawley and @charleso . Nice to see that we have two empirical tests for the runtime behavior; learning something new about the internals everyday :)

For completeness, the compile-time conversion behavior of type ascription is documented in the [Scala 2.11 specification](https://scala-lang.org/files/archive/spec/2.12/06-expressions.html#typed-expressions).

> The typed expression e:T has type T. The type of expression e is expected to conform to T. The result of the expression is the value of e converted to type T.